internal/clusterfeature/feature_manager.go

```diff
@@ -0,0 +1,316 @@
+// Copyright © 2019 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusterfeature
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/ghodss/yaml"
+	"github.com/goph/emperror"
+	"github.com/goph/logur"
+)
+
+const (
+	ExternalDnsChartVersion = "1.6.2"
+
+	ExternalDnsImageVersion = "v0.5.11"
+
+	ExternalDnsChartName = "stable/external-dns"
+
+	ExternalDnsNamespace = "default"
+
+	ExternalDnsRelease = "external-dns"
+)
+
+// ClusterService provides a thin access layer to clusters.
+type ClusterService interface {
+	// GetCluster retrieves the cluster representation based on the cluster identifier
+	GetCluster(ctx context.Context, clusterID uint) (Cluster, error)
+
+	// IsClusterReady checks whether the cluster is ready for features (eg.: exists and it's running).
+	IsClusterReady(ctx context.Context, clusterID uint) (bool, error)
+}
+
+// Cluster represents a Kubernetes cluster.
+type Cluster interface {
+	GetID() uint
+	GetOrganizationName() string
+	GetKubeConfig() ([]byte, error)
+}
+
+// externalDnsFeatureManager synchronous feature manager
+type externalDnsFeatureManager struct {
+	logger            logur.Logger
+	featureRepository FeatureRepository
+	clusterService    ClusterService
+	helmService       HelmService
+}
+
+// NewExternalDnsFeatureManager builds a new feature manager component
+func NewExternalDnsFeatureManager(logger logur.Logger, featureRepository FeatureRepository, clusterService ClusterService) FeatureManager {
+	hs := &featureHelmService{ // wired private component!
+		logger: logur.WithFields(logger, map[string]interface{}{"helm-service": "comp"}),
+	}
+	return &externalDnsFeatureManager{
+		logger:            logur.WithFields(logger, map[string]interface{}{"component": "feature-manager"}),
+		featureRepository: featureRepository,
+		clusterService:    clusterService,
+		helmService:       hs,
+	}
+}
+
+func (sfm *externalDnsFeatureManager) Activate(ctx context.Context, clusterId uint, feature Feature) error {
+	fLog := logur.WithFields(sfm.logger, map[string]interface{}{"cluster": clusterId, "feature": feature.Name})
+	fLog.Info("activating feature ...")
+
+	if _, err := sfm.featureRepository.GetFeature(ctx, clusterId, feature.Name); err != nil {
+		fLog.Debug("feature exists")
```

Comment:
When `err` is not `nil`, some kind of error occurred while getting the feature which doesn't mean the feature exists. Instead, the first return value should be checked for not being `nil`.

Reply:
nice catch, thanks